core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -209,7 +209,7 @@ trait Endpoint[A] { self =>
     final def apply(input: Input): Endpoint.Result[B] = self(input) match {
       case a @ EndpointResult.Matched(_, _) => other(input) match {
         case b @ EndpointResult.Matched(_, _) =>
-          if (a.rem.route.length <= b.rem.route.length) a else b
+          if (a.rem.route.length <= b.rem.route.length && a.rem.method == input.method) a else b
```

Comment:
So I've been thinking about this a lot and I'm not sure I like this particular piece. I tried experimenting with something in a local branch and it seems we can do something better. Here is an idea: encode method-not-allowed in `EndpointResult`. The scope of the change doesn't seem to break a lot of things:

The ADT would now have three cases to model HTTP routing result. It comes down to three things we're interested: a) route found b) route found but method is not allowed c) route not found

```scala
object EndpointResult {

  final case class Matched[A](rem: Input, out: Rerunnable[Output[A]]) extends EndpointResult[A] {
    def isMatched: Boolean = true
  }

  case object NotMatched extends EndpointResult[Nothing] { // previosuly Skipped
    def isMatched: Boolean = false
  }

  case object MethodNotMatched extends EndpointResult[Nothing] {
    def isMatched: Boolean = false
  }
}
```

The verb-matching combinators are now behaving correctly (returning an appropriate result):

```scala
      if (input.request.method == m) e(input)
      else e(input) match {
        case EndpointResult.Matched(_, _) => EndpointResult.MethodNotMatched
        case skipped => skipped
      }
```

Given the new endpoint result, `Endpoint.coproduct` can now be implemented more carefully. We don't have to check the method, we just need to return the most appropriate result. Here is the sketch:

```scala
  final def coproduct[B >: A](other: Endpoint[B]): Endpoint[B] = new Endpoint[B] {
    final def apply(input: Input): Endpoint.Result[B] = self(input) match {
      case a @ EndpointResult.Matched(_, _) => other(input) match {
        case b @ EndpointResult.Matched(_, _) =>
          if (a.rem.route.length <= b.rem.route.length) a else b
        case _ => a
      }
      case a => other(input) match {
        case EndpointResult.NotMatched => a
        case b => b
      }
    }
}
```

The rest is pretty much the same code as in this PR (think of changes to `ToService`).

@sergeykolbasov What do you think about this? I like how it's not breaking our `foo(Input.get("/"))`behavior.

Reply:
I experimented with it here: https://github.com/finagle/finch/pull/927